dali/operators/util/normal_distribution_op.cc

```diff
@@ -0,0 +1,103 @@
+where the shapes correspond to the shapes of tensors in the input batch.
+2. Providing a custom shape as an argument resolves in a batch of output tensors,
+where every output tensor has the same shape.
+3. Providing no input arguments resolves in a batch of output scalars,
+distributed normally (similar to `CoinFlip` operator).)code")
+                .NumInput(0, 1)
+                .NumOutput(detail::kNumOutputs)
+                .AddOptionalArg(detail::kMean, R"code(Mean value of the distribution)code",
+                                0.f, true)
+                .AddOptionalArg(detail::kStddev,
+                                R"code(Standard deviation of the distribution)code",
+                                1.f, true)
+                .AddOptionalArg(detail::kShape,
+                                R"code(Shape of single output tensor in a batch)code",
+                                detail::kShapeDefaultValue, true)
+                .AddOptionalArg(arg_names::kDtype, R"code(Data type for the output)code",
+                                DALI_FLOAT);
+
+DALI_REGISTER_OPERATOR(NormalDistribution, NormalDistributionCpu, CPU);
+
+
+#define NORM_TYPES (uint8_t, int8_t, uint16_t, int16_t, uint32_t, int32_t, uint64_t, int64_t, float, double)  // NOLINT
+
+
+bool NormalDistributionCpu::SetupImpl(std::vector<OutputDesc> &output_desc,
+                                      const workspace_t<CPUBackend> &ws) {
+  AcquireArguments(ws);
+  output_desc.resize(detail::kNumOutputs);
+  output_desc[0].shape = GetOutputShape(ws);
+  TYPE_SWITCH(dtype_, type2id, DType, NORM_TYPES, (
+          {
+            TypeInfo type;
+            type.SetType<DType>(dtype_);
+            output_desc[0].type = type;
+          }
+  ), DALI_FAIL(make_string("Unsupported output type: ", dtype_)))  // NOLINT
+  return true;
+}
+
+
+void NormalDistributionCpu::AssignSingleValueToOutput(workspace_t<CPUBackend> &ws) {
+  auto &output = ws.OutputRef<CPUBackend>(0);
+  TYPE_SWITCH(dtype_, type2id, DType, NORM_TYPES, (
+          distribution_t distribution(mean_[0], stddev_[0]);
+          for (int sample_id = 0; sample_id < batch_size_; ++sample_id) {
+              auto ptr = output[0].mutable_data<DType>();
+              *ptr = ConvertSat<DType>(distribution(rng_));
+          }
+  ), DALI_FAIL(make_string("Unsupported output type: ", dtype_)))  // NOLINT
+}
+
+
+void NormalDistributionCpu::AssignTensorToOutput(workspace_t<CPUBackend> &ws) {
+  auto &output = ws.OutputRef<CPUBackend>(0);
+  auto &tp = ws.GetThreadPool();
+  TYPE_SWITCH(dtype_, type2id, DType, NORM_TYPES, (
+            for (int sample_id = 0; sample_id < batch_size_; ++sample_id) {
+                distribution_t distribution(mean_[sample_id], stddev_[sample_id]);
+                auto ptr = output[sample_id].mutable_data<DType>();
+                for (long j = 0; j < volume(output[sample_id].shape()); j++) {  // NOLINT (long)
+                    ptr[j] = ConvertSat<DType>(distribution(rng_));
```

Comment:
I'm not sure if getting floats and saturating them to int is the the valid approach.

Reply:
My rationale here was as follows: User specifies the normal distribution he wants (by providing mean and stddev). If he provides `mean=0, stddev=1` and then provides `dtype=INT`, it's basically his fault, that he will get like 3 values here. But they will be distributed normally anyhow.

Again, if he provides `mean=999999 stddev=1, dtype=INT8`, it's also his fault, that he gets garbage.

On the other hand, this approach gives flexibility about the output type for the price that user has to know, what he's doing. I think I can assume that.

Reply:
I'd rather throw an error if the provided mean/stddev is to big for the given dtype

Reply:
What does it mean, that mean/stddev is too big for given dtype? e.g. I can imagine, that user provides mean=0 stddev=255 and dtype=UINT8 to create Bernoulli-ish distribution.

I guess the term GIGO would fit here...